### PR TITLE
Add compile functionality to cli

### LINKED ISSUE #636: Add compile functionality to cli
Imagine:
1. building a large flow with sub-flows and operations.

**Expected result:**
a. that you want to compile everything to avoid compile related errors at run-time.
b. that you want all the compile errors to be displayed not errors to appear one by one

**Actual result:** At run-time (using the run command) compile errors are triggered which are hard to differentiate among other errors and appear one by one.

Example:
`cslang>run --f /tutorials/base/print.sl --i text=Hi
Command failed java.lang.RuntimeException: java.lang.RuntimeException: For operation 'print' syntax is illegal.
Could not transform Input : {text=null} Since it has a null value.

Make sure a value is specified or that indentation is properly done.`

### PR DESCRIPTION
**Pull request by gituser173**
Fixes #636
Signed-off-by: Claudiu Ban claudiu-laurian.ban@hpe.com

### CODE REVIEW

**Comment by gituser173** [comment]
capture

File: cloudslang-all/src/main/java/io/cloudslang/lang/api/SlangImpl.java [hunk 1]
```diff
@@ -67,6 +68,20 @@ public CompilationArtifact compile(SlangSource source, Set<SlangSource> dependen
...
+    public CompilationModellingResult compileSource(SlangSource source, Set<SlangSource> dependencies) {
+
+        Validate.notNull(source, "Source can not be null");
+        Set<SlangSource> dependencySources = new HashSet<>(filter(notNullValue(), dependencies));
+
+        try {
+            return compiler.compileSource(source, dependencySources);
+        } catch (Exception e) {
+            logger.error("Failed compilation for source : " + source.getFileName() + " ,Exception is : " + e.getMessage());
+            throw new RuntimeException(e);
```
Comment (Bonczidai) [inline cloudslang-all/src/main/java/io/cloudslang/lang/api/SlangImpl.java]: shouldn't this be ~ new RE("same message as above", e) ?

File: cloudslang-cli/src/main/java/io/cloudslang/lang/cli/SlangCLI.java [hunk 2]
```diff
@@ -136,10 +136,37 @@ private boolean invalidVerboseInput(String verbose) {
...
+    private String printCompileErrors(List<RuntimeException> exceptions) {
+        StringBuilder stringBuilder = new StringBuilder();
+        if (exceptions.size() > 0) {
+            stringBuilder.append("Following exceptions were found:\n");
+            for (RuntimeException exception : exceptions) {
+                stringBuilder.append("\t");
+                stringBuilder.append(exception.getClass());
+                stringBuilder.append(": ");
+                stringBuilder.append(exception.getMessage());
+                stringBuilder.append("\n");
```
Comment (Bonczidai) [inline cloudslang-cli/src/main/java/io/cloudslang/lang/cli/SlangCLI.java]: WDYT about using File.separator here? Please check it on Linux

File: cloudslang-cli/src/main/java/io/cloudslang/lang/cli/SlangCLI.java [hunk 3]
```diff
@@ -136,10 +136,37 @@ private boolean invalidVerboseInput(String verbose) {
...
+            }
+            throw new RuntimeException(stringBuilder.toString());
+        } else {
+            stringBuilder.append("Compilation was successful");
```
Comment (Bonczidai) [inline cloudslang-cli/src/main/java/io/cloudslang/lang/cli/SlangCLI.java]: .. for file: abc.sl

-- maybe

File: cloudslang-cli/src/main/java/io/cloudslang/lang/cli/utils/CompilerHelperImpl.java [hunk 4]
```diff
@@ -63,13 +63,51 @@
...
     @Override
-    public CompilationArtifact compile(String filePath, List<String> dependencies) throws IOException {
+    public CompilationArtifact compile(String filePath, List<String> dependencies) {
+        File file = getFile(filePath);
+
+        try {
+            return slang.compile(SlangSource.fromFile(file), getDependencySources(dependencies, file));
+        } catch (Exception e) {
+            logger.error("Failed compilation for file : " + file.getName() + " ,Exception is : " + e.getMessage());
+            throw e;
```
Comment (Bonczidai) [inline cloudslang-cli/src/main/java/io/cloudslang/lang/cli/utils/CompilerHelperImpl.java]: same for wrapper exception here

File: cloudslang-cli/src/main/java/io/cloudslang/lang/cli/utils/CompilerHelperImpl.java [hunk 5]
```diff
@@ -63,13 +63,51 @@
...
+
+    @Override
+    public CompilationModellingResult compileSource(String filePath, List<String> dependencies) {
+        File file = getFile(filePath);
+
+        try {
+            return slang.compileSource(SlangSource.fromFile(file), getDependencySources(dependencies, file));
+        } catch (Exception e) {
+            logger.error("Failed compilation for file : " + file.getName() + " ,Exception is : " + e.getMessage());
+            throw e;
```
Comment (Bonczidai) [inline cloudslang-cli/src/main/java/io/cloudslang/lang/cli/utils/CompilerHelperImpl.java]: same for wrapper exception here

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/SlangCompiler.java [hunk 6]
```diff
@@ -29,6 +30,16 @@
     CompilationArtifact compile(SlangSource source, Set<SlangSource> path);
 
     /**
+     * Compile a CloudSlang source and its path to a {@link io.cloudslang.lang.entities.CompilationArtifact} object
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/SlangCompiler.java]: IMHO `dependencies` is better than `path`

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/SlangCompilerImpl.java [hunk 7]
```diff
@@ -68,25 +69,41 @@
...
 
                 executablePairs.put(preCompiledCurrentSource, currentSource);
             }
         }
 
-        executablePairs.remove(executable);
-        return scoreCompiler.compile(executable, executablePairs.keySet());
+        executablePairs.remove(executableModellingResult.getExecutable());
+
+        CompilationModellingResult result = scoreCompiler.compile(executableModellingResult.getExecutable(),
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/SlangCompilerImpl.java]: bit weird formatting

**Comment by Bonczidai** [comment]
and tests also for the CompilationModellingReuslt feature (errors are collected from scorecompile stage)

**Comment by Bonczidai** [comment]
IMHO adding support for `compile --d dir1,dir2` is a useful feature to compile everything from those folders

**Comment by lucian-cm** [comment]
@gituser173 waiting for build to pass, so that this can be merged

**Pull request merged**